// @midwayjs/core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FORMAT = exports.FileUtils = exports.PathFileUtil = exports.Types = exports.retryWith = exports.retryWithAsync = exports.extend = exports.Utils = exports.sleep = exports.isTypeScriptEnvironment = exports.wrapAsync = exports.wrapMiddleware = exports.pathMatching = exports.transformRequestObjectByType = exports.deprecatedOutput = exports.delegateTargetAllPrototypeMethod = exports.delegateTargetProperties = exports.delegateTargetMethod = exports.delegateTargetPrototypeMethod = exports.loadModule = exports.safeRequire = exports.safelyGet = exports.ASYNC_ROOT_CONTEXT = exports.MidwayPriorityManager = exports.DEFAULT_PRIORITY = exports.DataSourceManager = exports.WebRouterCollector = exports.MidwayServerlessFunctionService = exports.MidwayWebRouterService = exports.MidwayHealthService = exports.MidwayMockService = exports.MidwayDecoratorService = exports.MidwayMiddlewareService = exports.MidwayLifeCycleService = exports.MidwayAspectService = exports.MidwayFrameworkService = exports.MidwayLoggerService = exports.MidwayInformationService = exports.MidwayEnvironmentService = exports.MidwayConfigService = exports.FunctionalConfiguration = exports.createConfiguration = exports.BaseFramework = exports.MidwayRequestContainer = void 0;
__exportStar(require("./interface"), exports);
__exportStar(require("./context/container"), exports);
var requestContainer_1 = require("./context/requestContainer");
Object.defineProperty(exports, "MidwayRequestContainer", { enumerable: true, get: function () { return requestContainer_1.MidwayRequestContainer; } });
var baseFramework_1 = require("./baseFramework");
Object.defineProperty(exports, "BaseFramework", { enumerable: true, get: function () { return baseFramework_1.BaseFramework; } });
__exportStar(require("./context/providerWrapper"), exports);
__exportStar(require("./constants"), exports);
var configuration_1 = require("./functional/configuration");
Object.defineProperty(exports, "createConfiguration", { enumerable: true, get: function () { return configuration_1.createConfiguration; } });
Object.defineProperty(exports, "FunctionalConfiguration", { enumerable: true, get: function () { return configuration_1.FunctionalConfiguration; } });
var configService_1 = require("./service/configService");
Object.defineProperty(exports, "MidwayConfigService", { enumerable: true, get: function () { return configService_1.MidwayConfigService; } });
var environmentService_1 = require("./service/environmentService");
Object.defineProperty(exports, "MidwayEnvironmentService", { enumerable: true, get: function () { return environmentService_1.MidwayEnvironmentService; } });
var informationService_1 = require("./service/informationService");
Object.defineProperty(exports, "MidwayInformationService", { enumerable: true, get: function () { return informationService_1.MidwayInformationService; } });
var loggerService_1 = require("./service/loggerService");
Object.defineProperty(exports, "MidwayLoggerService", { enumerable: true, get: function () { return loggerService_1.MidwayLoggerService; } });
var frameworkService_1 = require("./service/frameworkService");
Object.defineProperty(exports, "MidwayFrameworkService", { enumerable: true, get: function () { return frameworkService_1.MidwayFrameworkService; } });
var aspectService_1 = require("./service/aspectService");
Object.defineProperty(exports, "MidwayAspectService", { enumerable: true, get: function () { return aspectService_1.MidwayAspectService; } });
var lifeCycleService_1 = require("./service/lifeCycleService");
Object.defineProperty(exports, "MidwayLifeCycleService", { enumerable: true, get: function () { return lifeCycleService_1.MidwayLifeCycleService; } });
var middlewareService_1 = require("./service/middlewareService");
Object.defineProperty(exports, "MidwayMiddlewareService", { enumerable: true, get: function () { return middlewareService_1.MidwayMiddlewareService; } });
var decoratorService_1 = require("./service/decoratorService");
Object.defineProperty(exports, "MidwayDecoratorService", { enumerable: true, get: function () { return decoratorService_1.MidwayDecoratorService; } });
var mockService_1 = require("./service/mockService");
Object.defineProperty(exports, "MidwayMockService", { enumerable: true, get: function () { return mockService_1.MidwayMockService; } });
var healthService_1 = require("./service/healthService");
Object.defineProperty(exports, "MidwayHealthService", { enumerable: true, get: function () { return healthService_1.MidwayHealthService; } });
var webRouterService_1 = require("./service/webRouterService");
Object.defineProperty(exports, "MidwayWebRouterService", { enumerable: true, get: function () { return webRouterService_1.MidwayWebRouterService; } });
var slsFunctionService_1 = require("./service/slsFunctionService");
Object.defineProperty(exports, "MidwayServerlessFunctionService", { enumerable: true, get: function () { return slsFunctionService_1.MidwayServerlessFunctionService; } });
Object.defineProperty(exports, "WebRouterCollector", { enumerable: true, get: function () { return slsFunctionService_1.WebRouterCollector; } });
var dataSourceManager_1 = require("./common/dataSourceManager");
Object.defineProperty(exports, "DataSourceManager", { enumerable: true, get: function () { return dataSourceManager_1.DataSourceManager; } });
var priorityManager_1 = require("./common/priorityManager");
Object.defineProperty(exports, "DEFAULT_PRIORITY", { enumerable: true, get: function () { return priorityManager_1.DEFAULT_PRIORITY; } });
Object.defineProperty(exports, "MidwayPriorityManager", { enumerable: true, get: function () { return priorityManager_1.MidwayPriorityManager; } });
__exportStar(require("./service/pipelineService"), exports);
__exportStar(require("./common/loggerFactory"), exports);
__exportStar(require("./common/serviceFactory"), exports);
__exportStar(require("./common/dataListener"), exports);
__exportStar(require("./common/fileDetector"), exports);
__exportStar(require("./common/webGenerator"), exports);
__exportStar(require("./common/middlewareManager"), exports);
__exportStar(require("./common/filterManager"), exports);
__exportStar(require("./common/applicationManager"), exports);
__exportStar(require("./setup"), exports);
__exportStar(require("./error"), exports);
var asyncContextManager_1 = require("./common/asyncContextManager");
Object.defineProperty(exports, "ASYNC_ROOT_CONTEXT", { enumerable: true, get: function () { return asyncContextManager_1.ASYNC_ROOT_CONTEXT; } });
// export decorator
__exportStar(require("./decorator"), exports);
__exportStar(require("./decorator/decoratorManager"), exports);
__exportStar(require("./decorator/constant"), exports);
// export utils
var util_1 = require("./util/");
Object.defineProperty(exports, "safelyGet", { enumerable: true, get: function () { return util_1.safelyGet; } });
Object.defineProperty(exports, "safeRequire", { enumerable: true, get: function () { return util_1.safeRequire; } });
Object.defineProperty(exports, "loadModule", { enumerable: true, get: function () { return util_1.loadModule; } });
Object.defineProperty(exports, "delegateTargetPrototypeMethod", { enumerable: true, get: function () { return util_1.delegateTargetPrototypeMethod; } });
Object.defineProperty(exports, "delegateTargetMethod", { enumerable: true, get: function () { return util_1.delegateTargetMethod; } });
Object.defineProperty(exports, "delegateTargetProperties", { enumerable: true, get: function () { return util_1.delegateTargetProperties; } });
Object.defineProperty(exports, "delegateTargetAllPrototypeMethod", { enumerable: true, get: function () { return util_1.delegateTargetAllPrototypeMethod; } });
Object.defineProperty(exports, "deprecatedOutput", { enumerable: true, get: function () { return util_1.deprecatedOutput; } });
Object.defineProperty(exports, "transformRequestObjectByType", { enumerable: true, get: function () { return util_1.transformRequestObjectByType; } });
Object.defineProperty(exports, "pathMatching", { enumerable: true, get: function () { return util_1.pathMatching; } });
Object.defineProperty(exports, "wrapMiddleware", { enumerable: true, get: function () { return util_1.wrapMiddleware; } });
Object.defineProperty(exports, "wrapAsync", { enumerable: true, get: function () { return util_1.wrapAsync; } });
Object.defineProperty(exports, "isTypeScriptEnvironment", { enumerable: true, get: function () { return util_1.isTypeScriptEnvironment; } });
Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return util_1.sleep; } });
Object.defineProperty(exports, "Utils", { enumerable: true, get: function () { return util_1.Utils; } });
var extend_1 = require("./util/extend");
Object.defineProperty(exports, "extend", { enumerable: true, get: function () { return extend_1.extend; } });
__exportStar(require("./util/webRouterParam"), exports);
__exportStar(require("./util/contextUtil"), exports);
__exportStar(require("./util/pathToRegexp"), exports);
__exportStar(require("./util/httpclient"), exports);
var retry_1 = require("./util/retry");
Object.defineProperty(exports, "retryWithAsync", { enumerable: true, get: function () { return retry_1.retryWithAsync; } });
Object.defineProperty(exports, "retryWith", { enumerable: true, get: function () { return retry_1.retryWith; } });
var types_1 = require("./util/types");
Object.defineProperty(exports, "Types", { enumerable: true, get: function () { return types_1.Types; } });
var pathFileUtil_1 = require("./util/pathFileUtil");
Object.defineProperty(exports, "PathFileUtil", { enumerable: true, get: function () { return pathFileUtil_1.PathFileUtil; } });
var fs_1 = require("./util/fs");
Object.defineProperty(exports, "FileUtils", { enumerable: true, get: function () { return fs_1.FileUtils; } });
var format_1 = require("./util/format");
Object.defineProperty(exports, "FORMAT", { enumerable: true, get: function () { return format_1.FORMAT; } });
